server/src/MessageHandler.ts

```diff
@@ -107,3 +187,100 @@ export async function handleDocumentFormatting(
     TextEdit.replace(fullDocumentRange(document), formatted),
   ])
 }
+
+/**
+ *
+ * This handler provides the initial list of the completion items.
+ */
+export function handleCompletionRequest(
+  params: CompletionParams,
+  documents: TextDocuments<TextDocument>,
+): CompletionList | undefined {
+  const context = params.context
+  if (context == null) {
+    return undefined
+  }
+
+  const document = documents.get(params.textDocument.uri)
+  if (!document) {
+    return undefined
+  }
+
+  const foundBlock = getBlockAtPosition(params.position.line, document)
+  if (!foundBlock) {
+    return getSuggestionForBlockTypes(document)
+  }
+
+  if (isFirstInsideBlock(params.position, document)) {
+    return getSuggestionForFirstInsideBlock(
+      foundBlock.type,
+      document,
+      params.position,
+      foundBlock,
+    )
+  }
+
+  // Completion was triggered by a triggerCharacter
+  if (context.triggerKind === 2) {
+    switch (context.triggerCharacter) {
+      case '@':
+        return getSuggestionsForAttributes(
+          foundBlock.type,
+          params.position,
+          document,
+        )
+      case '"':
+        return getSuggestionForSupportedFields(
+          foundBlock.type,
+          document,
+          params.position,
+        )
+    }
+  }
+
+  if (foundBlock.type === 'model') {
+    const symbolBeforePosition = document.getText({
+      start: {
+        line: params.position.line,
+        character: params.position.character - 1,
+      },
+      end: { line: params.position.line, character: params.position.character },
+    })
+    const currentLine = getCurrentLine(document, params.position.line).trim()
+    const wordsBeforePosition: string[] = currentLine
+      .substring(0, params.position.character - 1)
+      .trim()
+      .split(/\s+/)
+
+    if (currentLine.includes('(')) {
+      return getSuggestionsForInsideAttributes(
+        document,
+        params.position,
+        foundBlock,
+      )
+    }
+
+    // check if type
+    if (
+      wordsBeforePosition.length < 2 ||
+      (wordsBeforePosition.length === 2 && symbolBeforePosition != ' ')
+    ) {
+      return getSuggestionsForTypes(foundBlock, document)
+    }
+    return getSuggestionsForAttributes(
+      foundBlock.type,
+      params.position,
+      document,
+    )
+  }
+}
+
+/**
+ *
+ * @param item This handler resolves additional information for the item selected in the completion list.
+ */
+export function handleCompletionResolveRequest(
+  item: CompletionItem,
+): CompletionItem {
+  return item
```

Comment:
Is the a placeholder for something else in the future?

Reply:
No, that function is needed in order to show the docs during auto-completion.

Reply:
Ah interesting!